pkg/subscription/subscriber.go

```diff
@@ -0,0 +1,39 @@
+package subscription
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+// Subscriber is an interface to be implemented by components that can extract
+// relevant subscription details from a kargoapi.RepoSubscription and use them
+// to discover corresponding artifacts.
+type Subscriber interface {
+	// ApplySubscriptionDefaults applies default values to the provided
+	// kargoapi.RepoSubscription.
+	ApplySubscriptionDefaults(
+		ctx context.Context,
+		sub *kargoapi.RepoSubscription,
+	) error
+	// ValidateSubscription validates the provided kargoapi.RepoSubscription.
+	ValidateSubscription(
+		ctx context.Context,
+		f *field.Path,
+		sub kargoapi.RepoSubscription,
+	) field.ErrorList
```

Comment:
I will call out myself here that I don't entirely love this package having a dependency on apimachinery, however, since this method is implemented specifically so that a validating webhook can delegate some of its validation responsibilities to it and get back properly formatted errors, I think it makes a certain amount of sense to return field.Errorlist since that's seamlessly usable by the webhook.